File: OJ/Leetcode/Algorithm/34. Search for a Range.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        int first = 0, middle;
        int half, len;
        int size = nums.size();
        len = size;
    
        while(len > 0) {
            half = len >> 1;
            middle = first + half;
            if(nums[middle] < target) {     
                first = middle + 1;          
                len = len-half-1;       
            }
            else
                len = half;            
        }
        if(nums[first] != target)   first = -1;
        ans.push_back(first);
        first = 0;
        len = size;
        while(len > 0){
            half = len >> 1;
            middle = first + half;
            if(nums[middle] > target)     
                len = half;
            else{
                first = middle + 1;    
                len = len - half - 1;
            }
        }
        if(nums[first-1] != target)   first = 0;
        ans.push_back(first-1);
        return ans;
    }
};

File: OJ/Leetcode/Algorithm/34. Search for a Range.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        l = -1
        r = len(nums) - 1
        ans = [0, 0]
        while l+1 < r:
            mid = l + (r - l >> 1)
            if nums[mid] >= target:
                r = mid
            else:
                l = mid
        if nums[r]!= target:
            ans[0] = -1
        else:
            ans[0] = r
        
        l = 0
        r = len(nums)
        while l+1 < r:
            mid = l + (r - l >> 1)
            if nums[mid] <= target:
                l = mid
            else:
                r = mid
        
        if nums[l]!= target:
            ans[1] = -1
        else:
            ans[1] = l
        return ans
            